File: web/app/backend/ServiceFailover.scala
package backend

import logging.{CorrelatedLogging, CorrelationContext}
import models.consul.{HealthInfo, NodeInfo}
import play.api.libs.concurrent.Execution.Implicits._
import scaldi.{Injectable, Injector}

import scala.concurrent.Future
import scala.util.Random

class ServiceFailover(implicit inj: Injector) extends Injectable with CorrelatedLogging {
  private val consulLookup = inject[ConsulLookup]
  private val random = new Random(System.currentTimeMillis())

  def retry[R](serviceName: String, uri: String)(requester: String => Future[R])(implicit collectionContext: CorrelationContext): Future[R] = {
    def tryRequest(healthInfos: Seq[HealthInfo], offset:Int, idx: Int): Future[R] = {
      val healthInfo = healthInfos((offset + idx) % healthInfos.size)
      val url = s"http://${healthInfo.Node.Address}:${healthInfo.Service.Port}$uri"

      withMdc {
        log.info(s"Service try: $url")
      }
      requester(url).recoverWith {
        case e: Throwable =>
          withMdc {
            log.info(s"Trigger retry on $e")
          }
          if (idx + 1 >= healthInfos.size) {
            Future.failed(e)
          } else {
            tryRequest(healthInfos, offset, idx + 1)
          }
      }
    }

    consulLookup.lookup(serviceName).flatMap {
      healthInfos =>
        if (healthInfos.isEmpty) {
          Future.failed(new RuntimeException(s"No active service nodes for service $serviceName"))
        } else {
          val offset = random.nextInt(healthInfos.size)

          tryRequest(healthInfos, offset, 0)
        }
    }
  }
}

File: billing/src/main/java/de/leanovate/dose/billing/connector/ServiceFailover.java
package de.leanovate.dose.billing.connector;

import java.io.IOException;
import java.util.List;
import java.util.Random;

public class ServiceFailover {
    private static final Random RANDOM = new Random(System.currentTimeMillis());

    public static <E, T> T retry(final List<E> endpoints, final Requester<E, T> requester) throws IOException {

        final int size = endpoints.size();

        if (size == 0) {
            throw new RuntimeException("No active endpoints found");
        }

        final int offset = RANDOM.nextInt(size);
        IOException lastException = null;

        for (int idx = 0; idx < size; idx++) {
            final E endpoint = endpoints.get((idx + offset) % size);

            try {
                return requester.perform(endpoint);
            } catch (IOException e) {
                lastException = e;
            }
        }
        throw lastException;
    }

    @FunctionalInterface
    public interface Requester<E, T> {
        T perform(E endpoing) throws IOException;
    }
}
